### DISCUSSION TITLE
No way to synchronously flush async logger?

### DISCUSSION THREAD

**Discussion by sketch34** [gabime/spdlog#2203] [Q&A]
https://github.com/gabime/spdlog/wiki/7.-Flush-policy

I want to use a spdlog async logger in my `boost::assertion_failed_msg` implementation. `spdlog::shutdown()` is not appropriate in this case, I want logging to continue if debugging is continued.

```
namespace boost
{
    inline void assertion_failed(char const * expr, char const * function, char const * file, long line)
    {
      SPDLOG_ERROR("boost assert failed: '{}' in function {} @ {}:{}", expr, function, file, line);
      assert(false);
    }

    inline void assertion_failed_msg(char const * expr, char const * msg, char const * function, char const * file, long line)
    {
      SPDLOG_ERROR("boost assert msg failed: {}\n\t'{}' in function {} @ {}:{}", msg, expr, function, file, line);
      assert(false);
    }
} // namespace boost
```

I call `spdlog::flush_on()` appropriately, but it seems like there is no way to cause the async logger to flush to output before the program is halted by the `assert(false)`.

**Comment by gabime** [comment]
Right. This is a known issue with the async logger (https://github.com/gabime/spdlog/issues/1696). I would just put small sleep before the assert (ugly but effective).

**Comment by tt4g** [comment]
Another way, using `spdlog::apply_all()` (should use thread-safe sink).

```cpp
spdlog::apply_all([](std::shared_ptr<logger> logger)
{
    for (auto &sink : logger->sinks()) {
        sink->flush();
    }
});
```

**Comment by gabime** [comment]
> Another way, using spdlog::apply_all() (should use thread-safe sink

The problem it would be executed before the async queue processed fully.

**Reply by tt4g** [reply]
That's right.
Since synchronous communication is performed during the execution of asynchronous processing, timing problems will occur.
There is a risk that the flush operation will be completed before some messages are added to sink.
